### ISSUE TITLE
Unexpected behavior in jax.numpy.array_equal

### ISSUE THREAD

**Issue by jheek** [google/jax#17621]
### Description

The implementation of array_equal always returns False when the an input cannot be cast to an array using `jax.numpy.asarray`. This results in surprising equality checks when you accidentally feed an object:

```python
import dataclasses
@dataclasses.dataclass
class Foo:
  v: int

jax.numpy.array_equal(Foo(1), Foo(1)) # ==> False
jax.numpy.array_equal(Foo(1), Foo(2)) # ==> False
```

Normal numpy will revert to the __equals__ implementation of the object. I think this or simply throwing an error if the inputs are not castable to arrays would be reasonable behavior.

### What jax/jaxlib version are you using?

Any

### Which accelerator(s) are you using?

Any

**Comment by jakevdp** [comment]
Thanks for the report! See also this report: https://github.com/google/jax/issues/14901

I think the fix here would be to use `_check_arraylike` on the inputs, which would effectively raise an error if the inputs can't be converted to a JAX array. We haven't done this yet because it's a breaking change that requires a deprecation cycle.

**Comment by kfirdev** [comment]
I Think it might be better to fall-back to python equality but give out a warning about it, don't you think?

**Comment by jakevdp** [comment]
No, I don't think accepting non-array inputs in `array_equal` is a good idea, for the reasons mentioned in #7737

**Comment by jakevdp** [comment]
Falling back to Python equality would also lead to strange consequences; for example, what should the output of this be?
```python
jnp.array_equal(Foo(1), np.arange(4))
```
If we fall back to python `==`, we get this:
```python
>>> Foo(1) == np.arange(4)
array([False, False, False, False])
```
when a scalar boolean is expected.

**Comment by kfirdev** [comment]
Yes you are right, sorry.

**Comment by jakevdp** [comment]
Thanks for following up!
